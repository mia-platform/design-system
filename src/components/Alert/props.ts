/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

import { IconComponent } from '../Icon/Icon.props.ts'
import { Type } from './types.ts'

export type AlertProps = {

  /**
   * The title displayed at the top of the alert.
   */
  title?: ReactNode

  /**
   * The description providing additional details about the alert.
   */
  description?: ReactNode

  /**
   * Custom children content to be rendered inside the alert.
   */
  children?: ReactNode

  /**
   * Indicates if the alert is closable.
   */
  isClosable?: boolean

  /**
   * The type of alert.
   */
  type?: Type

  /**
   * If true, renders a compact version of the alert with reduced padding.
   */
  isCompressed?: boolean

  /**
   * A custom icon to display in the alert. Replaces the default icon based on the `type`.
   */
  icon?: IconComponent

  /**
   * A custom action element (e.g., a button) to display inside the alert.
   */
  action?: ReactNode

  /**
   * A callback function triggered when the alert is closed.
   */
  onClose?: () => void
}
