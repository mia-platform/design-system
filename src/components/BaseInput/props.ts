import { InputRef } from 'antd'
import { Ref } from 'react'

import { Appearance } from './types.ts'

export type BaseInputProps = {

  /**
   * The placeholder for the input.
   */
  placeholder?: string

  /**
   * The type of the Input.
   */
  appearance?: Appearance

  /**
   * Whether the input is disabled.
   */
  isDisabled?: boolean,

  /**
   * Whether the input has error.
   */
  isError?: boolean

  /**
   * Whether the input is read-only.
   */
  isReadOnly?: boolean

  /**
   * Whether the input is full width. Defaults to true.
   */
  isFullWidth?: boolean

  /**
   * Ref to the antd input.
   */
  inputRef?: Ref<InputRef>

  /**
   * Classes for the input.
   */
  className?: string

}
