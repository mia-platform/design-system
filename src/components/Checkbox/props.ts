/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { CheckboxChangeEvent } from 'antd/es/checkbox'
import { ReactNode } from 'react'

export type CheckboxProps = {

  /**
   * Additional description of the checkbox, to be rendered below the checkbox `text`.
   * Is ignored if the `text` prop is not provided.
   */
  description?: ReactNode,

  /**
   * Allows you to control whether the checkbox is checked.
   */
  isChecked?: boolean,

  /**
   * Allows you to control whether the checkbox is disabled. Defaults to false.
   */
  isDisabled?: boolean,

  /**
   * Allows you to control whether the checkbox is checked on its first render. Defaults to false.
   */
  isInitiallyChecked?: boolean,

  /**
   * Allows you to control whether the checkbox is indeterminate. Defaults to false.
   */
  isIndeterminate?: boolean

  /**
   * Function that is invoked when the checkbox state is changed.
   */
  onChange?: (event: CheckboxChangeEvent) => void

  /**
   * Text to be displayed next to the checkbox.
   */
  text?: ReactNode
}
