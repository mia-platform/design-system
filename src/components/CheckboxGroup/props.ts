/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
import { ReactNode } from 'react'

import { Direction } from './types.ts'

export type CheckboxGroupOption<T> = {

  /**
   * Adds the Radio id attribute value
   */
  value: T;

  /**
   * The text used to display the radio option
   */
  label: ReactNode;

  /**
   * description of the label that is placed underneath the label
   */
  description?: ReactNode;

  /**
   * Specifies whether the Radio option is disabled
   */
  disabled?: boolean;

}

export type CheckboxGroupProps<T> = {

  /**
   * The selected value displayed on first render
   */
  defaultValue?: T[];

  /**
   * Used for setting the currently selected value
   */
  value?: T[];

  /**
   * Options of the checkbox group
   */
  options: CheckboxGroupOption<T>[];

  /**
   * The callback function that is triggered when the state changes
   */
  onChange?: (checkedValue: T[]) => void;

  /**
   * If true, disable all checkboxes
   */
  isDisabled?: boolean;

  /**
   * The direction of the radio group. Defaults to 'vertical'
   */
  direction?: Direction
}
