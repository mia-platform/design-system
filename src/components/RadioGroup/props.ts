/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { MouseEventHandler, ReactNode } from 'react'
import { RadioChangeEvent } from 'antd'

import { Direction } from './types.ts'

export type RadioGroupOption<T> = {

  /**
   * Optional key associated to the option
   */
  key?: string

  /**
   * The value associated with the Radio option
   */
  value: T;

  /**
   * The text used to display the radio option
   */
  label: string;

  /**
   * description of the label that is placed underneath thelabel
   */
  description?: string;

  /**
   * Specifies whether the Radio option is disabled
   */
  disabled?: boolean;

  /**
   * Adds the Title attribute value
   */
  title?: string;

  /**
   * Adds the Radio Id attribute value
   */
  id?: string;
};

export type RadioGroupChangeEvent<T> = {

  /**
   * the value of the option selected
   */
  value: T;

  event: RadioChangeEvent;
};

export type RadioGroupProps<T> = {

  /**
   * Options of the group radio
   */
  options: RadioGroupOption<T>[];

  /**
   * The selected value displayed on first render
   */
  defaultValue?: T;

  /**
   * Used for setting the currently selected value
   */
  value?: T;

  /**
   * If true, disable all radio buttons
   */
  isDisabled?: boolean;

  /**
   * The callback function that is triggered when the state changes
   */
  onChange?: (changeEvent: RadioGroupChangeEvent<T>) => void;

  /**
   * The direction of the radio group. Defaults to 'vertical'
   */
  direction?: Direction
};

export type RadioProps<T> = {
  value?: T;
  children?: ReactNode
  className?: string
  label?: string;
  description?: string;
  isDisabled?: boolean;
  isChecked?: boolean;
  onClick?: MouseEventHandler<HTMLElement>;
  onChange?: (event: RadioChangeEvent) => void;
};
