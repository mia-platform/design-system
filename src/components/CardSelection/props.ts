/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

import { InputType, Layout } from './types.ts'
import { IconComponent } from '../Icon/Icon.props.ts'

export type CardSelectionOption<T> = {

  /**
   * Optional id value for the card.
   */
  id?: string

  /**
   * The main title displayed prominently on the card.
   */
  title: ReactNode

  /**
   * An optional icon displayed alongside the title.
   */
  icon?: IconComponent

  /**
   * A subtitle displayed below the title.
   */
  subtitle?: ReactNode

  /**
   * Custom content to be rendered inside the card.
   */
  content?: ReactNode

  /**
   * The value associated with the card.
   */
  value?: T

  /**
   * Determines if the card is disabled.
   */
  disabled?: boolean
}

export type CardSelectionProps<T> = {

  /**
   * The default value for the card selection.
   * Must be an array if inputType is "checkbox", a value if inputType is "radio".
   */
  defaultValue?: T | T[]

  gap?: number

  /**
   * The inputType of the card selection. Defaults: undefined (no input will be displayed).
   */
  inputType?: InputType,

  isDisabled?: boolean

  /**
   * The layout of the cards. Can be "horizontal" or "vertical". Default: "vertical".
   */
  layout: Layout

  /*
  * Handler triggered when the value of the selection changes (only triggered if inputType is "checkbox" or "radio")
  * */
  onChange?: (value: T | T[]) => void

  /*
  * Handler triggered when a card is clicked
  * */
  onClick?: (value?: T) => void

  /**
   * The options array for the card selection.
   */
  options: CardSelectionOption<T>[]

  /**
   * The controlled value for the card selection.
   * Must be an array if inputType is "checkbox", a value if inputType is "radio".
   */
  value?: T | T[]
}
