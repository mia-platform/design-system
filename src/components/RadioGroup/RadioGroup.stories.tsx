/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Meta, StoryObj } from '@storybook/react'

import { RadioGroupChangeEvent, RadioGroupOption, RadioGroupProps } from './props'
import { RadioGroup } from './RadioGroup'

const baseOptions: RadioGroupOption<number>[] = [
  {
    label: 'Option 1',
    value: 1,
  },
  {
    label: 'Option 2',
    value: 2,
  },
]

const optionsWithDescription: RadioGroupOption<number>[] = [
  {
    label: 'Option 1',
    value: 1,
    description: 'Description 1',
  },
  {
    label: 'Option 2',
    value: 2,
    description: 'Description 2',
  },
]

const optionsPartiallyDisabled: RadioGroupOption<number>[] = [
  {
    label: 'Option 1',
    value: 1,
    description: 'Description 1',
    disabled: true,
  },
  {
    label: 'Option 2',
    value: 2,
    description: 'Description 2',
  },
]

const allOptionsDisabled: RadioGroupOption<number>[] = [
  {
    label: 'Option 1',
    value: 1,
    description: 'Description 1',
    disabled: true,
  },
  {
    label: 'Option 2',
    value: 2,
    description: 'Description 2',
    disabled: true,
  },
]

const getArgs = <T, >(
  options: RadioGroupOption<T>[],
  defaultValue: T,
  disabled = false,
  onChange?: (changeEvent: RadioGroupChangeEvent<T>) => void
): RadioGroupProps<T> => ({
    options,
    defaultValue,
    isDisabled: disabled,
    onChange,
  })

const meta = {
  component: RadioGroup<number>,
  args: getArgs<number>(baseOptions, 1),
} satisfies Meta<typeof RadioGroup<number>>

export default meta

type Story = StoryObj<typeof meta>;

export const BasicExample: Story = {}

export const WithNonExistentDefaultValue: Story = {
  args: getArgs<number>(baseOptions, 3),
}

export const WithDescription: Story = {
  args: getArgs<number>(optionsWithDescription, 2),
}

export const Disabled: Story = {
  args: getArgs(baseOptions, 1, true),
}

export const PartiallyDisabled: Story = {
  args: getArgs(optionsPartiallyDisabled, 1),
}

export const AllOptionsDisabled: Story = {
  args: getArgs(allOptionsDisabled, 1),
}

const onChange: (changeEvent: RadioGroupChangeEvent<number>) => void = (event) => alert(`selected value is now ${event.value}`)
export const WithOnChange: Story = {
  args: getArgs(baseOptions, 1, false, onChange),
}

type MoreComplexType = { a: number; b: number };

const objectOption1: MoreComplexType = { a: 1, b: 2 }
const objectOption2: MoreComplexType = { a: 3, b: 4 }

const props: RadioGroupProps<MoreComplexType> = {
  options: [
    { value: objectOption1, label: 'option 1' },
    { value: objectOption2, label: 'option 2' },
  ],
  defaultValue: objectOption2,
  isDisabled: false,
}

const metaO = {
  component: RadioGroup<MoreComplexType>,
  args: props,
} satisfies Meta<typeof RadioGroup<MoreComplexType>>

type StoryO = StoryObj<typeof metaO>;

export const WithObjectValueType: StoryO = {
  args: props,
}
