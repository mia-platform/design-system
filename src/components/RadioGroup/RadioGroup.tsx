/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Radio as AntRadio, RadioChangeEvent } from 'antd'
import { ReactElement, ReactNode, useMemo } from 'react'

import { Radio } from './components/Radio'
import { RadioGroupProps } from './props'
import styles from './RadioGroup.module.css'

const { options: optionsClass } = styles

const defaults = {
  disabled: false,
}

export const RadioGroup = <T, >({
  defaultValue,
  value,
  isDisabled: disabled = defaults.disabled,
  options,
  onChange,
}: RadioGroupProps<T>): ReactElement => {
  const radioOptions = useMemo((): ReactNode => {
    return options.map((option, index) => (
      <Radio
        description={option.description}
        idDisabled={option.disabled}
        // eslint-disable-next-line react/no-array-index-key
        key={index}
        label={option.label}
        value={option.value}
      />
    ))
  }, [options])

  const handleChange = (event: RadioChangeEvent): void => {
    if (onChange) {
      onChange({ value: event.target.value, event })
    }
  }

  return (
    <AntRadio.Group
      defaultValue={defaultValue}
      disabled={disabled}
      onChange={handleChange}
      {...(value !== null && value !== undefined) ? { value } : {}}
    >
      <div className={optionsClass}>{radioOptions}</div>
    </AntRadio.Group>
  )
}
