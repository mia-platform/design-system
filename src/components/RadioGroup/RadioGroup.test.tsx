/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { render, screen, waitFor } from '@testing-library/react'

import { RadioGroupOption, RadioGroupProps } from './props'
import { RadioGroup } from './RadioGroup'
import { userEvent } from '../../test-utils'

const baseOptions: RadioGroupOption<number>[] = [
  {
    value: 1,
    label: 'option 1',
  },
  {
    value: 2,
    label: 'option 2',
  },
]

const optionsWithDescription: RadioGroupOption<number>[] = [
  {
    value: 1,
    label: 'option 1',
    description: 'description 1',
  },
  {
    value: 2,
    label: 'option 2',
    description: 'description 2',
  },
]

const partiallyDisabledOptions: RadioGroupOption<number>[] = [
  {
    value: 1,
    label: 'disabled option',
    disabled: true,
  },
  {
    value: 2,
    label: 'enabled option',
  },
  {
    value: 3,
    label: 'another enabled option',
  },
]

const baseProps: RadioGroupProps<number> = {
  defaultValue: 1,
  options: baseOptions,
}

describe('RadioGroup', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('renders correctly', () => {
    it('with base props', () => {
      const { asFragment } = render(<RadioGroup {...baseProps} />)
      expect(asFragment()).toMatchSnapshot()
    })

    it('with descriptions', () => {
      const { asFragment } = render(
        <RadioGroup {...{ ...baseProps, options: optionsWithDescription }} />
      )
      expect(asFragment()).toMatchSnapshot()
    })

    it('when disabled', () => {
      const { asFragment } = render(
        <RadioGroup {...{ ...baseProps, isDisabled: true }} />
      )
      expect(asFragment()).toMatchSnapshot()
    })

    it('with partially disabled options', () => {
      const { asFragment } = render(
        <RadioGroup {...{ ...baseProps, options: partiallyDisabledOptions }} />
      )
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('should invoke onChange with correct value on user selection', () => {
    it('with a number value type', async() => {
      const onChange = jest.fn()
      const props: RadioGroupProps<number> = {
        ...baseProps,
        onChange,
      }

      render(<RadioGroup {...props} />)

      const firstRadio = screen.getByRole('radio', { name: /option 1/ })
      const secondRadio = screen.getByRole('radio', { name: /option 2/ })

      expect(firstRadio).toBeChecked()
      expect(secondRadio).not.toBeChecked()

      userEvent.click(secondRadio)
      await waitFor(() => expect(onChange).toHaveBeenCalled())
      expect(firstRadio).not.toBeChecked()
      expect(secondRadio).toBeChecked()
      expect(onChange).toHaveBeenCalledTimes(1)
      const [[invocation1]] = onChange.mock.calls
      expect(invocation1.event.target.value).toEqual(2)
      expect(invocation1.value).toEqual(2)
    })

    it('with a string value type', async() => {
      const options: RadioGroupOption<string>[] = [
        {
          value: 'one',
          label: 'option 1',
        },
        {
          value: 'two',
          label: 'option 2',
        },
      ]

      const onChange = jest.fn()
      const props: RadioGroupProps<string> = {
        defaultValue: 'one',
        options,
        onChange,
      }

      render(<RadioGroup {...props} />)

      const firstRadio = screen.getByRole('radio', { name: /option 1/ })
      const secondRadio = screen.getByRole('radio', { name: /option 2/ })

      expect(firstRadio).toBeChecked()
      expect(secondRadio).not.toBeChecked()

      userEvent.click(secondRadio)
      await waitFor(() => expect(onChange).toHaveBeenCalled())
      expect(firstRadio).not.toBeChecked()
      expect(secondRadio).toBeChecked()
      expect(onChange).toHaveBeenCalledTimes(1)
      const [[invocation1]] = onChange.mock.calls
      expect(invocation1.event.target.value).toEqual('two')
      expect(invocation1.value).toEqual('two')
    })

    it('with a boolean value type', async() => {
      const options: RadioGroupOption<boolean>[] = [
        {
          value: true,
          label: 'option 1',
        },
        {
          value: false,
          label: 'option 2',
        },
      ]

      const onChange = jest.fn()
      const props: RadioGroupProps<boolean> = {
        defaultValue: true,
        options,
        onChange,
      }

      render(<RadioGroup {...props} />)

      const firstRadio = screen.getByRole('radio', { name: /option 1/ })
      const secondRadio = screen.getByRole('radio', { name: /option 2/ })

      expect(firstRadio).toBeChecked()
      expect(secondRadio).not.toBeChecked()

      userEvent.click(secondRadio)
      await waitFor(() => expect(onChange).toHaveBeenCalled())
      expect(firstRadio).not.toBeChecked()
      expect(secondRadio).toBeChecked()
      expect(onChange).toHaveBeenCalledTimes(1)
      const [[invocation1]] = onChange.mock.calls
      expect(invocation1.event.target.value).toEqual(false)
      expect(invocation1.value).toEqual(false)
    })

    it('with a object value type', async() => {
      type optionType = {a: number, b: number}
      const optionValue1 = { a: 1, b: 2 }
      const optionValue2 = { a: 3, b: 4 }
      const options = [{
        value: optionValue1,
        label: 'option 1',
      },
      {
        value: optionValue2,
        label: 'option 2',
      }]

      const onChange = jest.fn()
      const props: RadioGroupProps<optionType> = {
        defaultValue: optionValue1,
        options,
        onChange,
      }

      render(<RadioGroup {...props} />)

      const firstRadio = screen.getByRole('radio', { name: /option 1/ })
      const secondRadio = screen.getByRole('radio', { name: /option 2/ })

      expect(firstRadio).toBeChecked()
      expect(secondRadio).not.toBeChecked()

      userEvent.click(secondRadio)
      await waitFor(() => expect(onChange).toHaveBeenCalled())
      expect(firstRadio).not.toBeChecked()
      expect(secondRadio).toBeChecked()
      expect(onChange).toHaveBeenCalledTimes(1)
      const [[invocation1]] = onChange.mock.calls
      expect(invocation1.event.target.value).toEqual(optionValue2)
      expect(invocation1.value).toEqual(optionValue2)
    })
  })

  it('should not invoke onChange if an already selected option is clicked', async() => {
    const onChange = jest.fn()
    const props: RadioGroupProps<number> = {
      ...baseProps,
      onChange,
    }

    render(<RadioGroup {...props} />)

    const radio = screen.getByRole('radio', { name: /option 1/ })
    userEvent.click(radio)
    await waitFor(() => expect(onChange).not.toHaveBeenCalled())
  })

  it('should prevent selecting any option if the component is disabled', async() => {
    const onChange = jest.fn()
    const props: RadioGroupProps<number> = {
      ...baseProps,
      isDisabled: true,
      onChange,
    }

    render(<RadioGroup {...props} />)

    const firstRadio = screen.getByRole('radio', { name: /option 1/ })
    const secondRadio = screen.getByRole('radio', { name: /option 2/ })

    expect(firstRadio).toBeChecked()
    expect(secondRadio).not.toBeChecked()

    userEvent.click(firstRadio)
    await waitFor(() => expect(onChange).not.toHaveBeenCalled())
    userEvent.click(secondRadio)
    await waitFor(() => expect(onChange).not.toHaveBeenCalled())
    expect(firstRadio).toBeChecked()
    expect(secondRadio).not.toBeChecked()
  })

  it('should allow selection only of enabled options', async() => {
    const onChange = jest.fn()
    const props: RadioGroupProps<number> = {
      options: partiallyDisabledOptions,
      defaultValue: 2,
      onChange,
    }

    render(<RadioGroup {...props} />)

    const enabledOption = screen.getByRole('radio', {
      name: /another enabled option/,
    })
    const disabledOption = screen.getByRole('radio', {
      name: /disabled option/,
    })

    userEvent.click(enabledOption)
    await waitFor(() => expect(onChange).toHaveBeenCalled())
    expect(onChange).toHaveBeenCalledTimes(1)
    const [[invocation]] = onChange.mock.calls
    expect(invocation.value).toEqual(3)
    expect(invocation.event.target.value).toEqual(3)

    onChange.mockClear()

    userEvent.click(disabledOption)
    await waitFor(() => expect(onChange).not.toHaveBeenCalled())
  })
})
