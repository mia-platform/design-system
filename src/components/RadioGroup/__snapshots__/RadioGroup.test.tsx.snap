// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup renders correctly 1`] = `
<DocumentFragment>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-5wsri9"
  >
    <div
      class="options"
    >
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper ant-radio-wrapper-checked css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target ant-radio-checked"
            >
              <input
                checked=""
                class="ant-radio-input"
                type="radio"
                value="1"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  option 1
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target"
            >
              <input
                class="ant-radio-input"
                type="radio"
                value="2"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  option 2
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RadioGroup renders correctly when disabled 1`] = `
<DocumentFragment>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-5wsri9"
  >
    <div
      class="options"
    >
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper ant-radio-wrapper-checked ant-radio-wrapper-disabled css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target ant-radio-checked ant-radio-disabled"
            >
              <input
                checked=""
                class="ant-radio-input"
                disabled=""
                type="radio"
                value="1"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  option 1
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper ant-radio-wrapper-disabled css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target ant-radio-disabled"
            >
              <input
                class="ant-radio-input"
                disabled=""
                type="radio"
                value="2"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  option 2
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RadioGroup renders correctly with default value 1`] = `
<DocumentFragment>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-5wsri9"
  >
    <div
      class="options"
    >
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper ant-radio-wrapper-checked css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target ant-radio-checked"
            >
              <input
                checked=""
                class="ant-radio-input"
                type="radio"
                value="1"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  option 1
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target"
            >
              <input
                class="ant-radio-input"
                type="radio"
                value="2"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  option 2
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RadioGroup renders correctly with descriptions 1`] = `
<DocumentFragment>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-5wsri9"
  >
    <div
      class="options"
    >
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper ant-radio-wrapper-checked css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target ant-radio-checked"
            >
              <input
                checked=""
                class="ant-radio-input"
                type="radio"
                value="1"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  option 1
                </div>
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  description 1
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target"
            >
              <input
                class="ant-radio-input"
                type="radio"
                value="2"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  option 2
                </div>
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  description 2
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RadioGroup renders correctly with partially disabled options 1`] = `
<DocumentFragment>
  <div
    class="ant-radio-group ant-radio-group-outline css-dev-only-do-not-override-5wsri9"
  >
    <div
      class="options"
    >
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper ant-radio-wrapper-checked ant-radio-wrapper-disabled css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target ant-radio-checked ant-radio-disabled"
            >
              <input
                checked=""
                class="ant-radio-input"
                disabled=""
                type="radio"
                value="1"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  disabled option
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target"
            >
              <input
                class="ant-radio-input"
                type="radio"
                value="2"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  enabled option
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
      <div>
        <div
          class="radio"
        >
          <label
            class="ant-radio-wrapper css-dev-only-do-not-override-5wsri9"
          >
            <span
              class="ant-radio ant-wave-target"
            >
              <input
                class="ant-radio-input"
                type="radio"
                value="3"
              />
              <span
                class="ant-radio-inner"
              />
            </span>
            <span>
              <div
                class="radioContent"
              >
                <div
                  class="ant-typography bodyS css-dev-only-do-not-override-5wsri9"
                  role="paragraph"
                >
                  another enabled option
                </div>
              </div>
            </span>
          </label>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
